File: client/lib/auth/oidc_session.dart
/// `OidcSession` orchestrates the authorization-code-with-PKCE flow:
/// build the authorize URL, store the verifier in sessionStorage, and
/// (after redirect) exchange the code for tokens.
///
/// We dynamically discover the OP's authorization_endpoint and
/// token_endpoint via /.well-known/openid-configuration so dex / Auth0 /
/// Okta all work without per-OP code.
library;

import 'dart:async';
import 'dart:convert';

import 'package:http/http.dart' as http;

import '_session_storage_io.dart' if (dart.library.html) '_session_storage_html.dart' as ss;

import 'auth_state.dart';
import 'oidc_client.dart';

class OidcSession {
  final OidcConfig config;
  final AuthState authState;
  final http.Client httpClient;

  Map<String, dynamic>? _disco;
  Timer? _refreshTimer;

  OidcSession({
    required this.config,
    required this.authState,
    http.Client? httpClient,
  }) : httpClient = httpClient ?? http.Client();

  /// Discover (or return cached) OP metadata. Fetched once per session.
  Future<Map<String, dynamic>> discover() async {
    if (_disco != null) return _disco!;
    final url = '${config.issuer.replaceAll(RegExp(r"/$"), "")}/.well-known/openid-configuration';
    final resp = await httpClient.get(Uri.parse(url));
    if (resp.statusCode != 200) {
      throw StateError('oidc discovery failed: ${resp.statusCode}');
    }
    _disco = jsonDecode(resp.body) as Map<String, dynamic>;
    return _disco!;
  }

  /// Begin the login flow: stash a verifier in sessionStorage, then
  /// redirect the browser to the authorize URL. Never returns (on web).
  Future<void> beginLogin() async {
    final disco = await discover();
    final verifier = generateCodeVerifier();
    final challenge = codeChallengeS256(verifier);
    final state = generateState();

    ss.setItem('kitp_oidc_verifier', verifier);
    ss.setItem('kitp_oidc_state', state);

    final authEndpoint = disco['authorization_endpoint'] as String;
    final url = buildAuthorizeUri(
      authorizationEndpoint: authEndpoint,
      config: config,
      state: state,
      codeChallenge: challenge,
    );
    ss.assignLocation(url.toString());
  }

  /// Exchange the authorization code for tokens. Reads the verifier back
  /// out of sessionStorage; clears it on success.
  Future<void> handleCallback({required String code, required String state}) async {
    final storedState = ss.getItem('kitp_oidc_state');
    if (storedState == null || storedState != state) {
      throw StateError('oidc state mismatch');
    }
    final verifier = ss.getItem('kitp_oidc_verifier');
    if (verifier == null) {
      throw StateError('oidc verifier missing from sessionStorage');
    }
    final disco = await discover();
    final tokenEndpoint = disco['token_endpoint'] as String;
    final resp = await httpClient.post(
      Uri.parse(tokenEndpoint),
      headers: const {'Content-Type': 'application/x-www-form-urlencoded'},
      body: {
        'grant_type': 'authorization_code',
        'code': code,
        'redirect_uri': config.redirectUri,
        'client_id': config.clientId,
        'code_verifier': verifier,
      },
    );
    if (resp.statusCode != 200) {
      throw StateError('token exchange failed: ${resp.statusCode} ${resp.body}');
    }
    final body = jsonDecode(resp.body) as Map<String, dynamic>;
    _applyTokenResponse(body);
    // Cleanup the verifier — it's a one-shot value.
    ss.removeItem('kitp_oidc_verifier');
    ss.removeItem('kitp_oidc_state');
  }

  /// Use the refresh_token to rotate. Returns true on success.
  Future<bool> refresh() async {
    final rt = authState.refreshToken;
    if (rt == null || rt.isEmpty) return false;
    final disco = await discover();
    final tokenEndpoint = disco['token_endpoint'] as String;
    try {
      final resp = await httpClient.post(
        Uri.parse(tokenEndpoint),
        headers: const {'Content-Type': 'application/x-www-form-urlencoded'},
        body: {
          'grant_type': 'refresh_token',
          'refresh_token': rt,
          'client_id': config.clientId,
        },
      );
      if (resp.statusCode != 200) return false;
      final body = jsonDecode(resp.body) as Map<String, dynamic>;
      _applyTokenResponse(body);
      return true;
    } catch (_) {
      return false;
    }
  }

  void _applyTokenResponse(Map<String, dynamic> body) {
    final access = body['access_token'] as String?;
    final refresh = body['refresh_token'] as String?;
    final id = body['id_token'] as String?;
    final expiresIn = (body['expires_in'] as num?)?.toInt() ?? 600;
    final expiresAt = DateTime.now().add(Duration(seconds: expiresIn));
    if (access == null) {
      throw StateError('token response missing access_token');
    }
    authState.setTokens(
      accessToken: access,
      refreshToken: refresh,
      idToken: id,
      expiresAt: expiresAt,
    );
    _scheduleRefresh(expiresAt);
  }

  /// Schedule a refresh ~30 seconds before expiry. Cancels any prior timer.
  void _scheduleRefresh(DateTime expiresAt) {
    _refreshTimer?.cancel();
    final until = expiresAt.difference(DateTime.now()) - const Duration(seconds: 30);
    if (until.isNegative) return;
    _refreshTimer = Timer(until, () async {
      final ok = await refresh();
      if (!ok) authState.signOut();
    });
  }

  /// Cancel timers; call from app dispose.
  void dispose() {
    _refreshTimer?.cancel();
  }
}

File: client/src/auth/oidc_session.ts
// `OidcSession` orchestrates the authorization-code-with-PKCE flow:
// build the authorize URL, persist the verifier in sessionStorage, and
// (after redirect) exchange the code for tokens.
//
// Ports the runtime side of `client/lib/auth/oidc_session.dart`. The
// equivalent of the Dart conditional-import session storage shim is in
// `session_storage.ts`, which falls back to an in-memory map under tests.
//
// Discovery: the Dart implementation discovered authorization_endpoint /
// token_endpoint via `/.well-known/openid-configuration`. To keep the
// orchestrator easily testable (and to match the simpler endpoint shape
// kitpd's dex serves), we default to `${issuer}/auth` + `${issuer}/token`
// here. If a future OP needs different paths, a discovery layer can be
// added without touching the public API.

import type { OidcConfig, Tokens } from './oidc_client';
import {
  buildAuthorizeUrl,
  exchangeCode,
  generatePkce,
  generateState,
  refreshToken as refreshTokenCall,
} from './oidc_client';
import type { AuthState } from './auth_state.svelte';
import {
  clear as clearStorage,
  getVerifier,
  setVerifier,
  type MinimalStorage,
} from './session_storage';

type FetchFn = typeof fetch;

/// Pluggable navigation hook. Defaults to `window.location.assign(url)` in
/// the browser; tests pass a no-op or recorder.
type Navigate = (url: string) => void;

function defaultNavigate(url: string): void {
  const w = (globalThis as { location?: Location }).location;
  if (w && typeof w.assign === 'function') {
    w.assign(url);
  }
}

export interface OidcSessionOptions {
  storage?: MinimalStorage;
  fetchFn?: FetchFn;
  navigate?: Navigate;
}

export class OidcSession {
  private readonly config: OidcConfig;
  private readonly authState: AuthState;
  private readonly storage: MinimalStorage | undefined;
  private readonly fetchFn: FetchFn;
  private readonly navigate: Navigate;

  constructor(config: OidcConfig, authState: AuthState, opts: OidcSessionOptions = {}) {
    this.config = config;
    this.authState = authState;
    this.storage = opts.storage;
    this.fetchFn = opts.fetchFn ?? fetch;
    this.navigate = opts.navigate ?? defaultNavigate;
  }

  /// Begin the login flow: stash a verifier in sessionStorage, then
  /// redirect the browser to the authorize URL.
  async startSignIn(): Promise<void> {
    const { verifier, challenge } = await generatePkce();
    const state = generateState();
    if (this.storage) setVerifier(state, verifier, this.storage);
    else setVerifier(state, verifier);
    const url = buildAuthorizeUrl(this.config, state, challenge);
    this.navigate(url);
  }

  /// Handle the OP redirect back to redirect_uri. Validates state, looks
  /// the verifier back out of sessionStorage, posts to the token endpoint,
  /// and populates `authState`. Throws on state mismatch / missing
  /// verifier / token endpoint failure.
  async handleCallback(url: URL): Promise<void> {
    const code = url.searchParams.get('code');
    const state = url.searchParams.get('state');
    if (!code) throw new Error('oidc callback missing code');
    if (!state) throw new Error('oidc callback missing state');
    const verifier = this.storage
      ? getVerifier(state, this.storage)
      : getVerifier(state);
    if (verifier === null) {
      throw new Error('oidc state mismatch or verifier missing');
    }
    const tokens = await exchangeCode(this.config, code, verifier, this.fetchFn);
    this.authState.setTokens(tokens);
    if (this.storage) clearStorage(this.storage);
    else clearStorage();
  }

  /// Use the refresh_token to rotate. Returns true on success, false
  /// otherwise (the caller — typically the dispatcher's onUnauthorized
  /// hook — should sign out on false). Matches Dart `OidcSession.refresh`.
  async refreshIfNeeded(): Promise<boolean> {
    const rt = this.authState.refreshToken;
    if (rt === null || rt === '') {
      this.signOut();
      return false;
    }
    let tokens: Tokens;
    try {
      tokens = await refreshTokenCall(this.config, rt, this.fetchFn);
    } catch {
      this.signOut();
      return false;
    }
    this.authState.setTokens(tokens);
    return true;
  }

  /// Local sign-out: wipes tokens + clears the verifier from
  /// sessionStorage. Does NOT call the OP's end-session endpoint — that is
  /// a deliberate choice matching the Dart client.
  signOut(): void {
    this.authState.signOut();
    if (this.storage) clearStorage(this.storage);
    else clearStorage();
  }
}
